File: 0502-ipo/0502-ipo.java

class Solution {
    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        List<int[]> minCapital = new ArrayList<>();
        for (int i = 0; i < profits.length; i++) {
            minCapital.add(new int[]{capital[i], profits[i]});
        }
        Collections.sort(minCapital, (a, b) -> a[0] - b[0]);

        PriorityQueue<Integer> maxProfit = new PriorityQueue<>((a, b) -> b - a);
        int j = 0;
        for (int i = 0; i < k; i++) {
            while (j < minCapital.size() && minCapital.get(j)[0] <= w) {
                maxProfit.offer(minCapital.get(j)[1]);
                j++;
            }

            if (maxProfit.isEmpty())
                break;

            w += maxProfit.poll();
        }

        return w;
    }
}

File: 0502-ipo/0502-ipo.py
import heapq

class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: list, capital: list) -> int:
        projects = sorted(zip(capital, profits), key=lambda x: x[0])
        max_heap = []
        i = 0
        for _ in range(k):
            while i < len(projects) and projects[i][0] <= w:
                heapq.heappush(max_heap, -projects[i][1])
                i += 1
            if max_heap:
                w -= heapq.heappop(max_heap)
            else:
                break
        return w